### ISSUE TITLE
Dynamically creating arrays and objects

### ISSUE THREAD

**Issue by rogerl50** [ArnoldSmith86/virtualtabletop#545]
Currently, the only way to create a variable containing an array is something like
```
"var a = []",
"var a = push 'first'",
"var a = push 'second'",
...
```
until done. This is not particularly convenient. It seems odd that you can say
```
    {
      "func": "FOREACH",
      "in": [
        1,
        2,
        3,
        4,
        5,
        6
      ],
```
but that you can't easily define an array containing these numbers and then use the variable as `"in": "${array}"`.

I'm not sure there is any reasonable way to create a variable containing an object other than by having a property with the desired value and `GET`ting it.

**Comment by robartsd** [comment]
My first choice would bet to support "var a = [1,2,3,4,5,6]" in dynamic expressions, but that has some issues, especially if we want to support dynamic values within the array. We could easily support inline literal values with a new func.
```json
{
  "func":"VAR",
  "var":"a",
  "value": [
    1,
    2,
    3,
    4,
    5,
    6
  ]
}
```
Would this be more readable than using `"var a = ${PROPERTY varA}"` and storing the value in property `varA` of the widget?

**Comment by rogerl50** [comment]
Can you currently create an array property whose entries are computed and dynamic rather than just a static list? Completely ignoring questions of implementation and even language design, I'd like to see something like
```
"var a = '[ ${first}, ${second} ]'"
```
How do you create a property that is an array with those entries?

I agree, though, that just introducing a `VAR` func for static arrays is not better than putting the array in a property. For me the issue is more constructing an array on the fly.

**Comment by ArnoldSmith86** [comment]
> Where if the value in the object is an array we iterate over the elements expanding string templates. It would still convert the variables to strings.

Apart from the fact that there is no `VAR` operation, that last example would work now and it would not turn every element into a string if it contains **only** a variable.

**Comment by ArnoldSmith86** [comment]
And just to be clear, I'm not against supporting something like `var a = [ 1, 2, 3 ]`. It simply wasn't part of the initial PR. I would vote against non-empty object literals and nested arrays though. Those shouldn't be too hard either but I feel like it's a bit unwarranted. This is supposed to be a board game engine (and it's already possible; it just takes a few more lines)...

**Comment by bjalder26** [comment]
Personally, I kind of like the 'workaround' because it pulls the values out of the Routine.  So if I need to change the values later, then I don't need to hunt through the entire Routine.  However, with that being said, I kind of wish the properties were separated from the Routines in the JSON editor, because if some property falls between two Routines, it can be hard to find.

**Comment by larkob** [comment]
What if we could create arrays and objects within a simple JSON object in the routine? It doesn't even need a "func" property such as "VAR". We would simply interpret any property as a variable declaration, so you could set multiple variables if you wanted to.
```
"clickRoutine": [
...operations before....
   {
     "myVar1": [
        "a",
        "b"
      ],
     "myVar2": 3,
     "myBoolean": true,
     "someString": "foo",
     "myObject": {
       "foo": 1
     }
  },
...operations after....
]
```

**Comment by robartsd** [comment]
> What if we could create arrays and objects within a simple JSON object in the routine? It doesn't even need a "func" property such as "VAR". We would simply interpret any property as a variable declaration, so you could set multiple variables if you wanted to.
> 
> ```
> "clickRoutine": [
> ...operations before....
>    {
>      "myVar1": [
>         "a",
>         "b"
>       ],
>      "myVar2": 3,
>      "myBoolean": true,
>      "someString": "foo",
>      "myObject": {
>        "foo": 1
>      }
>   },
> ...operations after....
> ]
> ```

So any object without a "func" key would be interpreted as setting variables?

**Comment by larkob** [comment]
Yes, we could introduce the `"func": "VAR"` operation and use it as the default/fallback for objects in routines that did not specify any "func" (instead of responding with an error). As a regular operation it would look as follows:

```
"clickRoutine": [
   {
     "func": "VAR",
     "myArray": [
        "a",
        "b"
      ]
   }
]
```

The benefits of "VAR" as a regular operation would probably be documentation and contextual support in JSON editor.

**Comment by ArnoldSmith86** [comment]
I kinda like the idea. All the power of JSON without extra work for us.

**Comment by rogerl50** [comment]
What would you do with syntax such as
```
"var": 3,
"func": "abc"
```
The second one is a little clearer; the code would try to parse it as a `func` and fail, I guess. How about the first one? Do we want to be in the business of restricting name usage?

**Comment by ArnoldSmith86** [comment]
The order of the properties has to be considered random anyway. So the code can't (or shouldn't) rely on the order.

You could of course say `abc` is not a function we have, so put in the variable but I think it's just better in general to enforce `func: VAR`.

**Comment by rogerl50** [comment]
I thought the proposal was to allow `"myArray": [ "a", "b" ]` and to interpret it as if it had been written
```
   {
     "func": "VAR",
     "myArray": [
        "a",
        "b"
      ]
   }
```

**Comment by ArnoldSmith86** [comment]
I'm pretty sure something got lost in translation. So please add `clickRoutine` from now on.

This cannot work because `clickRoutine` is an array, not an object:

```
"clickRoutine": [
  "var x = 1",
  {
    "func": "SET"
  },
  "myArray": [ 1, 2 ]
]
```

This would work fine:

```
"clickRoutine": [
  "var x = 1",
  {
    "func": "SET"
  },
  {
    "func": "VAR",
    "myArray": [ 1, 2 ]
  }
]
```

Larkob suggested that `func` could be optional:

```
"clickRoutine": [
  "var x = 1",
  {
    "func": "SET"
  },
  {
    "myArray": [ 1, 2 ]
  }
]
```

**Comment by rogerl50** [comment]
Thanks for the clarification.

I would also prefer to have the explicit `"func": "VAR"`. But my real preference would be that instead of
```
"clickRoutine": [
  "var x = 1",
  {
    "func": "SET"
  },
  {
    "myArray": [ 1, 2 ]
  }
]
```
you write
```
"clickRoutine": [
  "var x = 1",
  {
    "func": "SET"
  },
  "var myArray = [ 1, 2 ]"
]
```
I gather from the discussion, though, that this syntax would still be restricted to literals, so that you could not write instead
```
"clickRoutine": [
  "var myArray = [ ${first}, ${second} ]"
]
```
I kind of regret having opened this can of worms to start with.

**Comment by ArnoldSmith86** [comment]
The "problem" with `"var myArray = [ 1, 2 ]"` is that we parse this in the engine. So every feature we want has to be added manually (so it could support variables in there, we would just have to code it.

The `"func": "SET"` approach uses the object and array structure of the JSON itself which means it would not require any code to support everything JSON can support.

**Comment by robartsd** [comment]
I'm also in favor of explicit `func:"VAR"` in the routine object. Do we expand references in strings? Which ones?

```json
"...Routine": [
  {
    "func":"VAR",
    "x": [
      "${y}",
      {
        "${z}":"${PROPERTY foo}"
      }
    ]
  }
```
Expanding references in any object keys would certainly require new code. I'm not sure how expansion currently works in nested objects. Also what would happen if a variable set is also referred to? I think all strings that are not object keys should be expanded first (recursively going into arrays and objects), then results should be assigned to variables based on the top level keys.

**Comment by ArnoldSmith86** [comment]
That already works. Variables are applied before anything of the operation is evaluated so you can't use the variables you're currently setting.

Accessing a variable named `${x}` shouldn't be a problem.

**Comment by robartsd** [comment]
> That already works. Variables are applied before anything of the operation is evaluated so you can't use the variables you're currently setting.
> 
> Accessing a variable named `${x}` shouldn't be a problem.

Other than accessing a variable with special characters in the name, that is what I thought; but I wasn't 100% sure and wanted to be explicit.

**Comment by gitran** [comment]
The VAR function was implemented and this can be closed?

**Comment by 96LawDawg** [comment]
Good catch.  Thank you.  Closed by #2276.
